src/User/UserLogin.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\User;
+
+use Yiisoft\Validator\Result;
+use Yiisoft\Validator\Rule\HasLength;
+use Yiisoft\Validator\Rule\Required;
+use Yiisoft\Validator\SimpleRuleHandlerContainer;
+use Yiisoft\Validator\Validator;
+
+final class UserLogin
+{
+    private function __construct(private string $value)
+    {
+    }
+
+    public function value(): string
+    {
+        return $this->value;
+    }
+
+    /**
+     * @throw UserLoginException
+     */
+    public static function createNew(string $login, UserRepository $userRepository): UserLogin
+    {
+        $validator = new Validator(new SimpleRuleHandlerContainer());
+        $rules = [
```

Comment:
Maybe move this validation into User ?

Reply:
UserLogin can't be in invalid. The validation should belong to him. You can't create invalid a login. 

You could use it for many cases, for example type hint in public setter methods of User, forms, console args and others.
User is orm model and you could use a schema typecast handler.

Reply:
Validation is not the same as asserting invariants. Validator wasn't meant for assertions. Also, checking for existing user is part of the business logic, not part of the invariant.